include/dlaf/communication/kernels/reduce.h

```diff
@@ -37,127 +34,194 @@ namespace comm {
-                const matrix::Tile<const T, D>&, MPI_Request* req) {
+template <class T>
+void reduceSend(comm::IndexT_MPI rank_root, const common::PromiseGuard<comm::Communicator>& pcomm,
+                MPI_Op reduce_op, const common::internal::ContiguousBufferHolder<const T>& cont_buf,
+                MPI_Request* req) {
   auto msg = comm::make_message(cont_buf.descriptor);
   auto& comm = pcomm.ref();
 
   DLAF_MPI_CALL(
       MPI_Ireduce(msg.data(), nullptr, msg.count(), msg.mpi_type(), reduce_op, rank_root, comm, req));
 }
 
-DLAF_MAKE_CALLABLE_OBJECT(reduceSend);
-
-}
-
-template <class T, Device D>
-void scheduleReduceRecvInPlace(const comm::Executor& ex,
-                               pika::future<common::PromiseGuard<comm::Communicator>> pcomm,
-                               MPI_Op reduce_op, pika::future<matrix::Tile<T, D>> tile) {
+// TODO SenderElementType of a reference_wrapper does not work
+template <class T, class Sender>
+auto senderReduceRecvInPlace(const comm::Executor& ex,
+                             pika::future<common::PromiseGuard<comm::Communicator>> pcomm,
+                             MPI_Op reduce_op, Sender&& tile) {
   // Note:
   //
-  // GPU  -> duplicateIfNeeded ---------------------> cCPU -> MPI -------------> copyIfNeeded --> GPU
-  // CPU  ----------------------> makeItContiguous -> cCPU -> MPI -> copyBack ------------------> CPU
-  // cCPU ------------------------------------------> cCPU -> MPI ------------------------------> cCPU
+  // CPU  ---> makeItContiguous ---> cCPU -> MPI ---> copyBack ---> CPU
+  // cCPU -------------------------> cCPU -> MPI -----------------> cCPU
   //
   // where: cCPU = contiguous CPU
 
-  using pika::dataflow;
-  using pika::unwrapping;
-
-  using common::internal::copyBack_o;
-  using common::internal::makeItContiguous_o;
-  using internal::reduceRecvInPlace_o;
-  using matrix::Tile;
-  using matrix::duplicateIfNeeded;
-  using matrix::getUnwrapReturnValue;
-  using matrix::unwrapExtendTiles;
-
-  const auto& ex_copy = getHpExecutor<Backend::MC>();
+  using namespace pika::execution::experimental;
+
+  using dlaf::matrix::Tile;
+  using dlaf::internal::Policy;
+  using dlaf::internal::transform;
+  using dlaf::internal::whenAllLift;
+
+  return std::forward<Sender>(tile) | transfer(thread_pool_scheduler{}) |
+         let_value([ex, reduce_op, pcomm = std::move(pcomm)](Tile<T, Device::CPU>& tile_orig) mutable {
+           using dlaf::common::internal::makeItContiguous;
+
+           return just(makeItContiguous(tile_orig)) |
+                  let_value([ex, reduce_op, &tile_orig,
+                             pcomm = std::move(pcomm)](const auto& cont_buffer) mutable {
+                    return pika::dataflow(ex, internal::reduceRecvInPlace<T>, pcomm.get(), reduce_op,
+                                          std::cref(cont_buffer)) |
+                           then([&]() {
```

Comment:
Keep in mind that this will be a synchronous continuation. May be intentional, but I just want to point it out. Feel free to close this without action if it's intentional. I guess `copyBack` is a relatively fast operation which might make sense to call synchronously.

Reply:
FIY @rasolca @teonnik and for the records.

It’s the copy back after the reduce, which must be completed as soon as possible because we can think of it as “part of the communication”. Previously (before senderification), we were requiring high priority just because of that.